src/main/drivers/compass/compass_lis3mdl.c

```diff
@@ -0,0 +1,203 @@
+#define LIS3MDL_DO_5                0x0C
+#define LIS3MDL_DO_10               0x10  // Default
+#define LIS3MDL_DO_20               0x14
+#define LIS3MDL_DO_40               0x18
+#define LIS3MDL_DO_80               0x1C
+#define LIS3MDL_FAST_ODR            0x02
+
+// CTRL_REG2
+#define LIS3MDL_FS_4GAUSS           0x00  // Default
+#define LIS3MDL_FS_8GAUSS           0x20
+#define LIS3MDL_FS_12GAUSS          0x40
+#define LIS3MDL_FS_16GAUSS          0x60
+#define LIS3MDL_REBOOT              0x08
+#define LIS3MDL_SOFT_RST            0x04
+
+// CTRL_REG3
+#define LIS3MDL_LP                  0x20  // Default 0
+#define LIS3MDL_SIM                 0x04  // Default 0
+#define LIS3MDL_MD_CONTINUOUS       0x00  // Default
+#define LIS3MDL_MD_SINGLE           0x01
+#define LIS3MDL_MD_POWERDOWN        0x03
+
+// CTRL_REG4
+#define LIS3MDL_ZOM_LP              0x00  // Default
+#define LIS3MDL_ZOM_MP              0x04
+#define LIS3MDL_ZOM_HP              0x08
+#define LIS3MDL_ZOM_UHP             0x0C
+#define LIS3MDL_BLE                 0x02  // Default 0
+
+// CTRL_REG5
+#define LIS3MDL_FAST_READ           0x80  // Default 0
+#define LIS3MDL_BDU                 0x40  // Default 0
+
+static bool lis3mdlRead(magDev_t * mag)
+{
+    bool ack;
+	uint8_t status = 0;
+
+	ack = busRead(mag->busDev, LIS3MDL_REG_STATUS_REG, &status);
+
+	if ( !( ack && ( ( status & 0x08 ) >> 3 ) ) ) {
+		return false;
+	}
+
+	uint8_t buf[6];
+
+    ack = busReadBuf(mag->busDev, LIS3MDL_REG_OUT_X_L, buf, 6);
+
+    if (!ack) {
+        mag->magADCRaw[X] = 0;
+        mag->magADCRaw[Y] = 0;
+        mag->magADCRaw[Z] = 0;
+        return false;
+    }
+
+    mag->magADCRaw[X] = (int16_t)(buf[1] << 8 | buf[0]) / 4;
+    mag->magADCRaw[Y] = (int16_t)(buf[3] << 8 | buf[2]) / 4;
+    mag->magADCRaw[Z] = (int16_t)(buf[5] << 8 | buf[4]) / 4;
+
+    return true;
+}
+
+#define DETECTION_MAX_RETRY_COUNT   200
+static bool deviceDetect(magDev_t * mag)
+{
+    for (int retryCount = 0; retryCount < DETECTION_MAX_RETRY_COUNT; retryCount++) {
+        delay(10);
+
+        uint8_t sig = 0;
+        bool ack = busRead(mag->busDev, LIS3MDL_REG_WHO_AM_I, &sig);
+
+        if (ack && sig == LIS3MDL_DEVICE_ID) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static bool lis3mdlInit(magDev_t *mag)
+{
+    delay(5);
+
+	busWrite(mag->busDev, LIS3MDL_REG_CTRL_REG2, LIS3MDL_FS_4GAUSS);   // Configuration Register 2
+    delay(5);
+
+    busWrite(mag->busDev, LIS3MDL_REG_CTRL_REG1, (LIS3MDL_TEMP_EN | LIS3MDL_OM_ULTRA_HI_PROF | LIS3MDL_DO_80) );   // Configuration Register 1
+    delay(5);
+
+    busWrite(mag->busDev, LIS3MDL_REG_CTRL_REG5, LIS3MDL_BDU);   // Configuration Register 5
+    delay(5);
+
+    busWrite(mag->busDev, LIS3MDL_REG_CTRL_REG4, LIS3MDL_ZOM_UHP);   // Configuration Register 4
+    delay(5);
+
+    busWrite(mag->busDev, LIS3MDL_REG_CTRL_REG3, 0x00);   // Configuration Register 3
+    delay(100);
+
+    return true;
+}
+
+bool lis3mdlDetect(magDev_t * mag)
```

Comment:
The I2C address of the LIS3MDL can be changed by connecting it's SDO/SA1 to either supply or ground.  As there is no way to know which configuration is used, it may be necessary to check both addresses.

Is it possible to change the I2C address of a device here, and then re-run the detection, or can that only be done in common_hardware.c?

Reply:
I'm planning to add support for this

Reply:
As I'm porting this over to Betaflight, I noticed that they have already added this ability. However, knowing that PX4 only supports the 0x1E, and since Betaflight is already having major space issues, I probably won't bother checking for the 0x1C address.